### DISCUSSION TITLE
Using a interface struct packet API

### DISCUSSION THREAD

**Discussion by Therzok** [NebulaModTeam/nebula#23] [Ideas]
You can avoid boxing of structs if you write packets using constrained generics: 

Something like the following definitions:
```csharp
interface IPacket
{
     void WriteTo(Stream s);
}

void Send<T>(T packet) where T:IPacket, struct
{
    packet.WriteTo(receiver);
}
```

This would completely avoid allocations or boxing of the struct if the constraints are kept wherever it's passed.

With unmanaged constraint you could write the whole struct in one go to the stream, but I'm not sure it's available.

It's possibly pretty important considering the impact of allocations might have on syncing.

**Comment by Mushroom** [comment]
Pull requests are welcome if you want to implement this :)

**Reply by Therzok** [reply]
I'll take a look this week. It might be easier than I initially thought. Are there any outstanding PRs or work that I should know of so I don't conflict?

**Reply by Mushroom** [reply]
@Therzok Yes, #45

**Comment by Therzok** [comment]
Looks like it's already been done in main
